src/Compilers/Core/Portable/Operations/Operation.cs

```diff
@@ -46,41 +76,203 @@ public Operation(OperationKind kind, SyntaxNode syntax, ITypeSymbol type, Option
 
         public abstract TResult Accept<TArgument, TResult>(OperationVisitor<TArgument, TResult> visitor, TArgument argument);
 
-        public static IOperation CreateOperationNone(SyntaxNode node, Optional<object> constantValue, Func<ImmutableArray<IOperation>> getChildren)
+        protected void SetParentOperation(IOperation parent)
         {
-            return new NoneOperation(node, constantValue, getChildren);
+            var result = Interlocked.CompareExchange(ref _parentDoNotAccessDirectly, parent, null);
+
+            // tree must belong to same semantic model if parent is given
+            Debug.Assert(parent == null || ((Operation)parent).SemanticModel == SemanticModel);
+
+            // make sure given parent and one we already have is same if we have one already
+            Debug.Assert(result == null || result == parent);
         }
 
-        private class NoneOperation : IOperation
+        public static IOperation CreateOperationNone(SemanticModel semanticModel, SyntaxNode node, Optional<object> constantValue, Func<ImmutableArray<IOperation>> getChildren)
         {
-            private readonly Func<ImmutableArray<IOperation>> _getChildren;
+            return new NoneOperation(semanticModel, node, constantValue, getChildren);
+        }
 
-            public NoneOperation(SyntaxNode node, Optional<object> constantValue, Func<ImmutableArray<IOperation>> getChildren)
+        public static T SetParentOperation<T>(T operation, IOperation parent) where T : IOperation
+        {
+            // operation can be null
+            if (operation == null)
             {
-                Syntax = node;
-                ConstantValue = constantValue;
-                _getChildren = getChildren;
+                return operation;
+            }
+
+            // explicit cast is not allowed, so using "as" instead
+            (operation as Operation).SetParentOperation(parent);
+            return operation;
+        }
+
+        public static ImmutableArray<T> SetParentOperation<T>(ImmutableArray<T> operations, IOperation parent) where T : IOperation
+        {
+            // check quick bail out case first
+            if (operations.Length == 0)
+            {
+                // no element
+                return operations;
+            }
+
+            // race is okay. penalty is going through a loop one more time
+            // explicit cast is not allowed, so using "as" instead
+            // invalid expression can have null element in the array
+            if ((operations[0] as Operation)?._parentDoNotAccessDirectly != null)
+            {
+                // already initialized
+                return operations;
```

Comment:
yep, that race can happen, but in those case, SearchParentPoint will set right parent to the field. so data will be still correct. it just takes slower route (searching parent operation) than cheaper route (parent being explicitly set by operation operation)